linux/ibus-keyman/src/engine.c

```diff
@@ -238,32 +253,39 @@ static void reset_context(IBusEngine *engine)
 }
 
 static void
-ibus_keyman_engine_init (IBusKeymanEngine *keyman)
-{
-    keyman->status_prop = ibus_property_new ("status",
-                                           PROP_TYPE_NORMAL,
-                                           NULL,
-                                           NULL,
-                                           NULL,
-                                           TRUE,
-                                           FALSE,
-                                           0,
-                                           NULL);
-    g_object_ref_sink(keyman->status_prop);
-    keyman->prop_list = ibus_prop_list_new ();
-    g_object_ref_sink(keyman->prop_list);
-    ibus_prop_list_append (keyman->prop_list,  keyman->status_prop);
-
-    keyman->table = ibus_lookup_table_new (9, 0, TRUE, TRUE);
-    g_object_ref_sink(keyman->table);
-    keyman->state = NULL;
+ibus_keyman_engine_init(IBusKeymanEngine *keyman) {
+  gdk_init(NULL, NULL);
+  keyman->status_prop = ibus_property_new("status", PROP_TYPE_NORMAL, NULL, NULL, NULL, TRUE, FALSE, 0, NULL);
+  g_object_ref_sink(keyman->status_prop);
+  keyman->prop_list = ibus_prop_list_new();
+  g_object_ref_sink(keyman->prop_list);
+  ibus_prop_list_append(keyman->prop_list, keyman->status_prop);
+
+  keyman->table = ibus_lookup_table_new(9, 0, TRUE, TRUE);
+  g_object_ref_sink(keyman->table);
+  keyman->state = NULL;
+  keyman->display = gdk_display_get_default();
+#ifdef GDK_WINDOWING_X11
+  keyman->xdisplay = NULL;
+  if (GDK_IS_X11_DISPLAY(keyman->display)) {
+    keyman->xdisplay = GDK_DISPLAY_XDISPLAY(keyman->display);
+  } else
+#endif
+#ifdef GDK_WINDOWING_WAYLAND
+  if (GDK_IS_WAYLAND_DISPLAY(keyman->display)) {
+  } else
+#endif
+  {
+    g_error("unsupported GDK backend");
+  }
```

Comment:
I take it this somewhat awkward construct is quite deliberate -- to allow for any permutation of `#define GDK_WINDOWING_X11` and `GDK_WINDOWING_WAYLAND` (with X11 taking precedence over Wayland if both are defined)?